### DISCUSSION TITLE
Use proxy as a fallback when version is missing in verdaccio

### DISCUSSION THREAD

**Discussion by ndelangen** [verdaccio/verdaccio#3516] [Ideas]
The storybook team has been using verdaccio in our CI for years now!

We have a feature request.

Current situation:
We setup verdaccio with a list of packages we want hosted by verdaccio, and everthing else gets proxied to the regular registry.
When we do this, we run into the following problem:
We publish a single version of a package to verdaccio, this is great. In fact we do this for about ~80 packages.
All these packages now appear on verdaccio, but they all have exactly 1 version on it.

The problem:
When a request comes in for a package that is in verdaccio, but with a different version. this fails. 
The verdaccio cache does not contain that version of that package.

The proposal:
Could we somehow configure verdaccio, so when a request for a package+version comes in, it checks locally if that package+version exists locally, and if not, proxy it to NPM.

In short:
Can we make verdaccio proxy packages that verdaccio knows about, but have that version missing?

**Comment by juanpicado** [comment]
> When a request comes in for a package that is in verdaccio, but with a different version. this fails.
The verdaccio cache does not contain that version of that package.

So you have like this right?

```
packages:
  '@storybook/*':
    access: $all
    publish: $authenticated
  '@*/*':
    access: $all
    publish: $authenticated
    proxy: npmjs
  '**':
    publish: $authenticated
    proxy: npmjs

**Reply by ndelangen** [reply]
Pretty much

https://github.com/storybookjs/storybook/blob/5a557d37ce573e15290cd7ed43ba434e3b42e6b5/scripts/verdaccio.yaml#L1-L133

**Reply by juanpicado** [reply]
I'm still thinking on it ... 🐌 matter of fact, I've been busy migrating to Storybook 6 😆

**Reply by ndelangen** [reply]
@juanpicado is your storybook project OSS?

**Reply by juanpicado** [reply]
Unfortunately no, but you have really good docs :-) so you have been helping me indirectly. It is just had long time without reading storybook docs and takes some time to migrate.

**Reply by juanpicado** [reply]
Well, now we have one :) https://ui-components.verdaccio.org/

**Comment by juanpicado** [comment]
> When a request comes in for a package that is in verdaccio, but with a different version. this fails.
> The verdaccio cache does not contain that version of that package.

```
packages:
  '@storybook/*':
    access: $all
    publish: $authenticated
    proxy_only: ['^1.x']
  '@*/*':
    access: $all
    publish: $authenticated
    proxy: npmjs
  '**':
    publish: $authenticated
    proxy: npmjs
```

I was thinking about this, what about a new property named `proxy_only` with semver support? Does this would help? 

I haven't check technically how hard might be just an idea.

**Comment by ndelangen** [comment]
That would be a bit inconvenient @juanpicado
...reason being: we'd have to track all the exact versions we're publishing. That's relatively easy right now, they are all the same, but that's something we want to change in the future.

How about: 
```
packages:
  '@storybook/*':
    access: $all
    publish: $authenticated
    fallback_proxy: npmjs
```

Which instructs verdaccio to proxy to npm, only when it can't find the requested package+version locally
